hax/hax/motr/delivery.py

```diff
@@ -69,14 +69,45 @@ def wait_for_any(self,
                       self.recently_delivered[promise])
             del self.recently_delivered[promise]
 
+    def wait_for_all(self,
+                     promise: HaLinkMessagePromise,
+                     timeout_sec: float = 30.0):
+        """
+        Blocks the current thread until all of the messages in
+        promise._ids are reported by Motr as delivered.
+
+        Raises NotDelivered exception when timeout_sec exceeds.
+        """
+        for msg in promise._ids:
+            condition = Condition()
+            with self.lock:
+                self.waiting_clients[promise] = condition
+
+            with condition:
+                LOG.debug('Blocking until %s is confirmed', promise)
+                condition.wait(timeout=timeout_sec)
+            with self.lock:
+                if promise not in self.recently_delivered:
+                    raise NotDelivered('None of message tags =' +
+                                       str(promise) +
+                                       '  were delivered to Motr')
+                msgid = self.recently_delivered[promise]
+                if msgid in promise:
+                    promise._ids.remove(msgid)
+                LOG.debug('Thread unblocked - %s just received',
+                          self.recently_delivered[promise])
+                del self.recently_delivered[promise]
```

Comment:
...but deletion of the promise must happen only when it is the last iteration

Otherwise there is potentially a time window when this promise is removed from the dict but a delivery happens.

Reply:
@mssawant I took a liberty to correct this code a bit. Please have a look at https://github.com/mssawant/cortx-hare-1/commit/34f2594b5e37e8cbf1a1143dfc377aca161181a5

I think without that dicts `self.recently_delivered` and `self.waiting_clients` will grow till infinity.

Reply:
Thanks, I see one more improvement while testing. Will update the patch.